### ISSUE TITLE
[question] Logins through lightdm fail trying to write to .xsession-errors in an encrypted home folder

### ISSUE THREAD

**Issue by jnvsor** [google/fscrypt#203]
Debian Sid, fscrypt `0.2.5-2`. Not sure if this is lightdm's fault or fscrypt's fault, but it's above my paygrade either way.

Attempting to log in through lightdm fails with the following in syslog:

```
lightdm[1945]: Failed to open log file .xsession-errors: Required key not available
```

After login and a brief black screen you get kicked back to lightdm. Further research finds the offending line to be in `/etc/X11/Xsession`:

```
exec >>"$ERRFILE" 2>&1
```

This results in errno 2. This still occurs if the folder is already unlocked via a login in another TTY. Strangely enough, you can still stat the file and write to other (0777) files in the home folder from the script:

```
echo "Start test" > "$HOME/debug"
echo "$ERRFILE" >> "$HOME/debug"
stat "$ERRFILE" >> "$HOME/debug"
set +e -x
echo "test" >> "$ERRFILE"
echo $? >> "$HOME/debug"
echo "test" | tee -a "$ERRFILE" 2>>"$HOME/debug"
echo $? >> "$HOME/debug"
set +x -e
id >> "$HOME/debug"
```

This results in the following output:

```
Start test
/home/j/.xsession-errors
  File: /home/j/.xsession-errors
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 805h/2053d	Inode: 88080688    Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/       j)   Gid: ( 1000/       j)
Access: 2020-03-02 21:40:36.249347457 +0100
Modify: 2020-03-02 21:40:36.249347457 +0100
Change: 2020-03-02 21:40:36.249347457 +0100
 Birth: -
2
tee: /home/j/.xsession-errors: Required key not available
1
uid=1000(j) gid=1000(j) groups=1000(j),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),108(netdev),113(lpadmin),117(scanner)
```

Session will start normally through `startx` in a TTY and produce this output:

```
Start test
/home/j/.xsession-errors
  File: /home/j/.xsession-errors
  Size: 12766     	Blocks: 32         IO Block: 4096   regular file
Device: 805h/2053d	Inode: 88080394    Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/       j)   Gid: ( 1000/       j)
Access: 2020-03-02 21:26:00.726821993 +0100
Modify: 2020-03-02 21:26:00.726821993 +0100
Change: 2020-03-02 21:26:00.726821993 +0100
 Birth: -
0
0
uid=1000(j) gid=1000(j) groups=1000(j),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),108(netdev),113(lpadmin),117(scanner)
```

~~Creating a new folder with a v2 policy had no effect, but~~ using an unencrypted home folder worked (Given the right permissions)

Clarification: 0.2.5 doesn't have v2 support, but the debian package 0.2.5-2 didn't produce the following error after making a second policy and encrypting a folder with it, so I presume it was backported:

> fscrypt (pid 4464) is setting deprecated v1 encryption policy; recommend upgrading to v2.

**Comment by ebiggers** [comment]
Using a v2 encryption policy will fix this.  Note that it requires `fscrypt` v0.2.6 or later along with kernel v5.4 or later.  Older versions of `fscrypt` ignore the `policy_version` setting in `/etc/fscrypt.conf` and always use v1.

**Comment by jnvsor** [comment]
@ebiggers Thanks, then I'll assume the debian version backported the option but not the implementation (Hence the v1/v2 error message)

I'll reopen this issue if once 0.2.6 lands in debian it doesn't fix the issue

**Comment by ebiggers** [comment]
I doubt that Debian has backported anything.  Note that the message "fscrypt (pid 4464) is setting deprecated v1 encryption policy; recommend upgrading to v2." is printed by the kernel, and is only printed once per boot.  So I think you were actually using v1 policies in both cases.  The recommended way to check which policy version a directory is using is via `fscrypt status $dir`.

Has the maintainer of the Debian package been informed that it's outdated?  Also, feel free to build  v0.2.6 yourself and use it.

**Comment by jnvsor** [comment]
>  Note that the message "fscrypt (pid 4464) is setting deprecated v1 encryption policy; recommend upgrading to v2." is printed by the kernel, and is only printed once per boot

That explains everything, thanks!

>  The recommended way to check which policy version a directory is using is via `fscrypt status $dir`

I tried that, but there was no version information. Presumably because I was on 0.2.5. I'd consider building it if it bothered me, but I honestly don't care for lightdm anyway and will happily use `startx` for a while.

The performance boost of ext4 encryption over a >10 year old ecryptfs is staggering. `du -sh .local` is 200ms now, I killed it after 2 minutes on the old one. Overall performance has seen similar sharp rises
